### ISSUE TITLE
equivalent gensim command compared to word2vec

### ISSUE THREAD

**Issue by simonzack** [piskvorky/gensim#426]
What is the equivalent of 

    word2vec -train sentences.txt -output sentence_vecs.txt -cbow 0 -size 100 -window 10 -negative 5 -hs 0 -sample 1e-4 -threads 40 -binary 0 -iter 20 -min-count 1 -sentence-vectors 1

in gensim? I've tried:

    Doc2Vec(dm=0, size=100, window=10, negative=5, hs=0, sample=1e-4, workers=8, iter=20, min_count=1)

but the resulting vectors are very small (with the mean absolute value being of `0.00251142` compared to word2vecs `0.153490513481`), and is no better than picking the most frequent class in a classification benchmark.

**Comment by piskvorky** [comment]
Let's ping @gojomo on this one, he knows doc2vec the best.

**Comment by gojomo** [comment]
Your Doc2Vec instantiation uses comparable settings. How are you feeding the data in? By not including the corpus in the instantiation, you'll have to both `model.build_vocab(corpus)` then `model.train(corpus)`. (The call to `train()` should respect the previously-set 'iter'.)

Those methods don't expect the same format as Mikolov's sentence-vectors demo code, which expects an ID as the first token in a space-delimited example-per-line text file. They instead expect an iterable of objects (such as the example TaggedDocument class) which have both a 'tags' property (a list of tags, but for ideal memory behavior with large datasets, the list should contain just a single ascending int) and a 'words' property (list of string tokens). 

A common mistake is to supply the untokenized document string as 'words' – so the individual letters get interpreted as the words, essentially noise. That might account for the low magnitudes you see; the other telltale would be a `len(model.index2word)` of just dozens rather than tens/hundreds-of-thousands.

**Comment by piskvorky** [comment]
Alternatively, can you [copy&paste](https://gist.github.com/) your log (DEBUG or INFO level) @simonzack ?

**Comment by simonzack** [comment]
Thanks @gojomo for the hints. There was an iter problem but fixing that still didn't move doc2vec above the baseline. I've put my code up [here](https://gist.github.com/simonzack/f828519516af19c31a25), the first version of fit is gensim and the second word2vec.

gensim log: https://gist.github.com/simonzack/c97347c5d08fee117939
word2vec log: https://gist.github.com/simonzack/c43b6242582456935ef4

**Comment by gojomo** [comment]
The gensim log suggests the expected parameters are in effect and training is happening. So, I suspect if you are getting worse results from Doc2Vec than word2vec.c+sentence-vectors, on a downstream task, it's the subsequent steps that are the issue. (I only see the creation of vectors in your code – not training of some other model for classification or evaluation of that model.)

From the example counts, it seems you might be using the "Stanford Sentiment Treebank" (rotten-tomatoes movie review) dataset? If so, beware that is a very thin dataset from which many have found it hard to get good results. (The results from the original Paragraph Vectors paper seem to be unreproducible; Mikolov's sentence-vectors demo script skips right to the IMDB dataset, which includes 10x the examples and 100x the words.) 

Still, if you're getting better results from word2vec.c+sentence-vectors than Doc2Vec, that's a discrepancy I'd like to resolve – the predictiveness of the resulting vectors from similar parameters should be similar. So some general thoughts on your setup:

* When I was doing similar experiments, I sometimes had bugs where my labels were mismatched with the doc vectors, or I was retrieving the wrong vector for an intended doc, and I've seen others make similar errors – so I would double-check for such an issue. 

* With a dataset this small, you might as well use string tags as your doc IDs – the same tokens as in the sentence-vectors case – for clarity of comparison. 

* I found that on small-dataset sentiment-prediction tasks, almost any subsampling (via the 'sample' parameter) decreased the doc vectors' predictiveness... though it's clear that in other word2vec tasks subsampling can sometimes both speed training *and* improve representations. 

* Mikolov's sentence-vectors patch mixes the DBOW training of the special, line-start doc-vector tokens with simultaneous skip-gram word-vector training. By default, Doc2Vec won't do that, because it's not strictly necessary to get a good doc-vector, and in fact it both slows training and may make the doc-vectors less powerful for some purposes (at least in my small sentiment tests). But you can get that behavior by adding 'dbow_words=1' to the configuration – and some other interesting results, such as those in the 'Document Embeddings with Paragraph Vectors' paper, do rely on such co-trained DBOW doc-vecs and word-vecs.

**Comment by simonzack** [comment]
I debugged through word2vec & gensim a bit so the code roughly matches up, and the following fixes things, thanks for the pointers!

    dm=0, dbow_words=1, size=100, window=10, negative=5, hs=0, sample=0, min_count=1, iter=20, workers=64
